web-ui/github.ml

```diff
@@ -337,24 +337,19 @@ let repo_handle ~meth ~owner ~name ~repo path =
       | Aborted | Failed _ | Passed | Undefined _ -> Lwt.return None
     in
     let cancel_many (commit: Client.Commit.t) (job_infos : Client.job_info list) =
-      let open Lwt.Infix in
-      Lwt_list.filter_map_p (fun job_info -> can_cancel commit job_info) job_infos >>= fun l ->
-        Lwt_list.map_p (fun (ji, j) -> Current_rpc.Job.cancel j |> Lwt_result.map @@ fun () -> ji) l >>= fun x ->
-          let success = ref [] in
-          let failed = ref 0 in
-          let log_error r =
-          match r with
-          | Ok(ji) -> success := ji :: !success
-          | Error (`Capnp ex) ->
-            incr failed;
-            Log.err (fun f -> f "Error cancelling job: %a" Capnp_rpc.Error.pp ex)
-          in
-          let _ = List.map log_error x in
-          Lwt.return (!success, !failed)
+      Lwt_list.fold_left_s (fun (success, failed) job_info ->
+        can_cancel commit job_info >>= function
+        | None -> Lwt.return (success, failed)
+        | Some (ji, j) ->
+            Current_rpc.Job.cancel j >|= function
+            | Ok () -> (ji :: success, failed)
+            | Error (`Capnp ex) ->
+                Log.err (fun f -> f "Error cancelling job: %a" Capnp_rpc.Error.pp ex);
+                (success, succ failed)
+      ) ([], 0) job_infos
```

Comment:
This part of the code was previously going through the list of jobs 3 times (Lwt_list.filter_map_p, Lwt_list.map_p, List.map) for not much. In opam-repo-ci this list of jobs can be extremely long (500_000 or more) so this part of the code should be really focused on optimizing such list operations.

Furthermore using the parallel operations (Lwt_list.map_p) might overwhelm the ocluster server with too many operations at once, so I changed it to a sequential one instead (Lwt_list.fold_left_s)

Reply:
It's a shame there's no `Lwt_list.iter_n` to do batches, like there is for `Lwt_stream`.

Reply:
🤦🏾‍♂️